Branch-SDK/src/io/branch/indexing/ContentDiscoverer.java

```diff
@@ -0,0 +1,256 @@
+
+    //------------------------- Public methods---------------------------------//
+
+    public void discoverContent(final Activity activity, String referredUrl) {
+        cdManifest_ = ContentDiscoveryManifest.getInstance(activity);
+        referredUrl_ = referredUrl;
+
+        //Scan for content only if the app is started by  a link click or if the content path for this view is already cached
+        ContentDiscoveryManifest.CDPathProperties pathProperties = cdManifest_.getCDPathProperties(activity);
+
+        if (pathProperties != null) { // Check if view is available in CD manifest
+            // Discover content only if element array is not empty json array
+            if (!pathProperties.isSkipContentDiscovery()) {
+                discoverContent(activity);
+            }
+        } else if (!TextUtils.isEmpty(referredUrl_)) {
+            discoverContent(activity);
+        }
+    }
+
+    private void discoverContent(Activity activity) {
+        if (discoveredViewList_.size() < cdManifest_.getMaxViewHistorySize()) { // check if max discovery views reached
+            handler_.removeCallbacks(readContentRunnable);
+            lastActivityReference_ = new WeakReference<>(activity);
+            handler_.postDelayed(readContentRunnable, VIEW_SETTLE_TIME);
+        }
+    }
+
+
+    public void onActivityStopped(Activity activity) {
+        if (lastActivityReference_ != null && lastActivityReference_.get() != null
+                && lastActivityReference_.get().getClass().getName().equals(activity.getClass().getName())) {
+            handler_.removeCallbacks(readContentRunnable);
+            lastActivityReference_ = null;
+        }
+    }
+
+    public void onSessionStarted(final Activity activity, String referredUrl) {
+        discoveredViewList_ = new ArrayList<>();
+        discoverContent(activity, referredUrl);
+    }
+
+    // ---------------Private methods----------------------//
+
+    private Runnable readContentRunnable = new Runnable() {
+        @Override
+        public void run() {
+
+            try {
+                if (cdManifest_.isCDEnabled() && lastActivityReference_ != null && lastActivityReference_.get() != null) {
+                    Activity activity = lastActivityReference_.get();
+                    JSONObject contentEvent = new JSONObject();
+                    contentEvent.put(TIME_STAMP_KEY, System.currentTimeMillis());
+                    contentEvent.put(NAV_PATH_KEY, contentNavPath_);
+                    if (!TextUtils.isEmpty(referredUrl_)) {
+                        contentEvent.put(REFERRAL_LINK_KEY, referredUrl_);
+                    }
+                    String viewName = "/" + activity.getClass().getSimpleName();
+                    contentEvent.put(VIEW_KEY, viewName);
+
+
+                    ViewGroup rootView = (ViewGroup) activity.findViewById(android.R.id.content);
```

Comment:
Perhaps null check if rootView is still present in the below. Since we have a weak reference to the activity, it could possibly be cleaned up prior to our referencing it?